### PR TITLE
Add string overloads to DbDataReader.Get*()

### PR DESCRIPTION
**Pull request by TheBlueSky**
Implement #31595

### CODE REVIEW

File: src/System.Data.Common/tests/System/Data/Common/DbDataReaderTest.cs [hunk 1]
```diff
@@ -28,113 +28,470 @@ namespace System.Data.Tests.Common
...
+
+            Assert.Equal(expected.Length, length);
+            for (int i = 0; i < expected.Length; i++)
+            {
+                Assert.Equal(expected[i], actual[i]);
+            }
+        }
+
+        [Fact]
+        public void GetCharByColumnNameTest()
```
Comment (Wraith2) [inline src/System.Data.Common/tests/System/Data/Common/DbDataReaderTest.cs]: This is going to throw on SqlClient because it doesn't support [`GetChar`](https://github.com/dotnet/corefx/blob/f26fb9c2a6fe52a5cdb9309f1b7dbe14e1e667b0/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs#L1844). I don't know what behaviour odbc has.
I don't think the test is wrong but it may be worth investigating if this test is run as part of the SqlClient tests and if so whether that behaviour needs to be taken into account somehow.

Comment (TheBlueSky) [reply src/System.Data.Common/tests/System/Data/Common/DbDataReaderTest.cs]: The test itself runs using a fake DB Reader (`DbDataReaderMock`). It should not impact or be impacted by `SqlDataReader`. Also, `SqlDataReader` will throw the same exception with `GetChar(string name)` that it throws with `GetChar(int i)`; which it the expected behaviour.

Unless I misunderstood your comment or misunderstood how things work :)

**Comment by TheBlueSky** [comment]
@roji, any hint regarding the failing checks? :|

File: src/System.Data.Common/tests/System/Data/Common/DbDataReaderTest.cs (hunk 1 above)
Comment (Wraith2) [reply src/System.Data.Common/tests/System/Data/Common/DbDataReaderTest.cs]: Ah, the Common mislead me. There's some stuff in src\Common that gets included as files into other projects. My bad. Looks good.

**Comment by Wraith2** [comment]
I think you can ignore the non-ci builds, those are being replaced by the CI versions.
There are actual build failures though,
```
2019-02-09T18:44:29.0849657Z System\Data\Common\DbDataReaderTest.cs(88,49): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1079635Z System\Data\Common\DbDataReaderTest.cs(103,46): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1111832Z System\Data\Common\DbDataReaderTest.cs(119,47): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1176328Z System\Data\Common\DbDataReaderTest.cs(138,46): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1181986Z System\Data\Common\DbDataReaderTest.cs(156,47): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1370569Z System\Data\Common\DbDataReaderTest.cs(175,50): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1371002Z System\Data\Common\DbDataReaderTest.cs(190,49): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1371297Z System\Data\Common\DbDataReaderTest.cs(205,48): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1371582Z System\Data\Common\DbDataReaderTest.cs(251,60): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1371838Z System\Data\Common\DbDataReaderTest.cs(270,47): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1372129Z System\Data\Common\DbDataReaderTest.cs(285,46): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1372418Z System\Data\Common\DbDataReaderTest.cs(300,47): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1372696Z System\Data\Common\DbDataReaderTest.cs(315,47): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1373154Z System\Data\Common\DbDataReaderTest.cs(330,47): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1373875Z System\Data\Common\DbDataReaderTest.cs(379,48): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1374228Z System\Data\Common\DbDataReaderTest.cs(402,48): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1374599Z System\Data\Common\DbDataReaderTest.cs(451,56): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1374942Z System\Data\Common\DbDataReaderTest.cs(469,47): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1375318Z System\Data\Common\DbDataReaderTest.cs(483,47): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
2019-02-09T18:44:29.1375689Z System\Data\Common\DbDataReaderTest.cs(484,46): error CS1503: Argument 1: cannot convert from 'string' to 'int' [D:\a\1\s\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
```
try doing `build src\System.Data.Common -allconfigurations` which will run all the various builds and you'll get errors from some of them. You may have to dabble with the conditional inclusions in the csproj file to include/exclude some things.

**Comment by TheBlueSky** [comment]
Yeah, so those error in the logs, but couldn't figure out why. It looks like the rest class couldn't see the changes in `src/System/Data/Common/DbDataReader.cs` for some reason.

Would like a hint on where to start looking.

**Comment by Wraith2** [comment]
Does it repro locally for the failing target? So if you specify the netfx target do you get a clean build locally?

**Comment by TheBlueSky** [comment]
`build src\System.Data.Common -allconfigurations` didn't work with me; I got the following error:

```
C:\Users\Essam\.nuget\packages\microsoft.dotnet.corefxtesting\1.0.0-beta.19105.4\build\core\Core.targets(56,5): error MSB3030: Could not copy the file "E:\Projects\.github\corefx\artifacts\bin\runtime/netcoreapp-Windows_NT-Debug-x64/xunit.execution.desktop.dll" because it was not found. [E:\Projects\.github\corefx\src\System.Data.Common\tests\System.Data.Common.Tests.csproj]
```

And then the errors that are shown in the CI builds; i.e., `error CS1503: Argument 1: cannot convert from 'string' to 'int'`.

However, when I try to build without specifying the framework or build individual frameworks, the build succeeds; e.g., `build -framework netcoreapp`, `build -framework netfx`, and `build -framework uap`.

**Comment by Wraith2** [comment]
Have you done `build -allconfigurations` and `build -release` from the corefx root? those two should restore all the dependencies.

**Comment by TheBlueSky** [comment]
I ran `build -allconfigurations` and `build -release` and both completed successfully. However, I'm still getting the error shown in the CI builds (not the "Could not copy the file" one) when I ran `build src\System.Data.Common -allconfigurations`.

I followed this document (https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md) to get started. I'm not sure I missed something.

Also, if the build fails in CI pipeline, I'm not sure it's related to dependencies. For some reason, certain configurations cannot see the changes I did to `src/System/Data/Common/DbDataReader.cs` file.

**Comment by Wraith2** [comment]
It's not just you, I've replicated the problem and I know my setup works. Started from clean (`clean -all`) and then ran the allconfigurations and release builds as I asked you to. Then branched and edited in your changes. The tests don't build once they're applied. 

What I can see is how the tests are supposed to be resolving the main library, there's no reference in the project. If it's using nuget then it won't work until the package is rebuilt and I've no good way to do that (when required I edited the nuget package but obviously that won't work in the ci).

So, bat signal time @safern to the rescue! hopefully...

**Comment by Wraith2** [comment]
Lets see if the area owners can help, /cc @divega, @ajcvickers, @afsanehr, @david-engel, @tarikulsabbir

**Comment by divega** [comment]
@karelz Any idea of can help with this? Are checks working normally at the moment? When I click on the Details for the failed checks I get 404s.

**Comment by safern** [comment]
The thing is that some of the builds doesn’t exist anymore, they where deleted. So we need to close and reopen the PR.

**Comment by safern** [comment]
Also it would be great if this PR could be rebased on top of master and push a new commit so that the CI system reboots and the latest changes take effect.

**Comment by TheBlueSky** [comment]
@safern, I rebased the branch and the build results are accessible and show the error.

**Comment by TheBlueSky** [comment]
@roji, can you have a look?

File: src/System.Data.Common/src/System/Data/Common/DbDataReader.cs [hunk 2]
```diff
@@ -61,40 +65,73 @@ public virtual DataTable GetSchemaTable()
...
+        public DbDataReader GetData(string name) => GetData(GetOrdinal(name));
+
         IDataReader IDataRecord.GetData(int ordinal) => GetDbDataReader(ordinal);
 
         protected virtual DbDataReader GetDbDataReader(int ordinal)
         {
             throw ADP.NotSupported();
         }
 
+        public DbDataReader GetDbDataReader(string name) => GetDbDataReader(GetOrdinal(name));
```
Comment (bgrainger) [inline src/System.Data.Common/src/System/Data/Common/DbDataReader.cs]: This should not be implemented; delete this line.

https://github.com/dotnet/corefx/issues/31595#issuecomment-459019657

File: src/System.Data.Common/ref/System.Data.Common.cs [hunk 3]
```diff
@@ -1945,43 +1945,71 @@ public abstract partial class DbDataReader : System.MarshalByRefObject, System.C
...
         public abstract long GetBytes(int ordinal, long dataOffset, byte[] buffer, int bufferOffset, int length);
+        public long GetBytes(string name, long dataOffset, byte[] buffer, int bufferOffset, int length) => throw null;
         public abstract char GetChar(int ordinal);
+        public char GetChar(string name) => throw null;
         public abstract long GetChars(int ordinal, long dataOffset, char[] buffer, int bufferOffset, int length);
+        public long GetChars(string name, long dataOffset, char[] buffer, int bufferOffset, int length) => throw null;
         [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
         public System.Data.Common.DbDataReader GetData(int ordinal) { throw null; }
+        [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
+        public System.Data.Common.DbDataReader GetData(string name) { throw null; }
```
Comment (bgrainger) [inline src/System.Data.Common/ref/System.Data.Common.cs]: Nitpick: Use `{ throw null; }` or `=> throw null;` consistently in this file; don't use both syntaxes.

**Comment by benaadams** [comment]
Does `System.Data.Common` type forward to inbox on NETFX and UWP?

i.e. do they need to be netcoreapp only?

File: src/System.Data.Common/ref/System.Data.Common.cs [hunk 4]
```diff
@@ -1945,43 +1945,71 @@ public abstract partial class DbDataReader : System.MarshalByRefObject, System.C
...
         public abstract string GetDataTypeName(int ordinal);
+        public string GetDataTypeName(string name) => throw null;
         public abstract System.DateTime GetDateTime(int ordinal);
+        public System.DateTime GetDateTime(string name) => throw null;
         protected virtual System.Data.Common.DbDataReader GetDbDataReader(int ordinal) { throw null; }
+        public System.Data.Common.DbDataReader GetDbDataReader(string name) { throw null; }
```
Comment (bgrainger) [inline src/System.Data.Common/ref/System.Data.Common.cs]: This should not be added to the API.

Comment (TheBlueSky) [reply src/System.Data.Common/ref/System.Data.Common.cs]: Remove it

File: src/System.Data.Common/src/System/Data/Common/DbDataReader.cs (hunk 2 above)
Comment (TheBlueSky) [reply src/System.Data.Common/src/System/Data/Common/DbDataReader.cs]: Removed it

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 3 above)
Comment (TheBlueSky) [reply src/System.Data.Common/ref/System.Data.Common.cs]: Agree. In my defence though, it was already inconsistent 😄

Anyway, Change it eveywhere in `DbDataReader` to `=> throw null;`.

**Comment by Wraith2** [comment]
It looks like @benaadams might have found the answer. All corefx assemblies have an inbox attribute applied through metadata `[assembly: AssemblyMetadata("PreferInbox", "True")]` which the loader will check and then revert to the inbox version if there is one. So on netfx since there's a version of System.Data.Common already it loads that. 

The fix is to make the ref class partial and use a second file called `System.Data.Common.netcoreapp.cs` to hold the new parts and then in the csproj conditionally include that element using a condition like `Condition="'$(TargetsNetCoreApp)' == 'true'`. You'll also have to make the tests netcoreapp only or they'll fail to compile or fail to run.

I could be wrong but it's worth a try I'd say, the only other option is to keep waiting.

File: src/System.Data.Common/ref/System.Data.Common.cs [hunk 5]
```diff
@@ -1939,58 +1939,85 @@ public abstract partial class DbDataReader : System.MarshalByRefObject, System.C
         public abstract object this[int ordinal] { get; }
         public abstract object this[string name] { get; }
         public abstract int RecordsAffected { get; }
-        public virtual int VisibleFieldCount { get { throw null; } }
+        public virtual int VisibleFieldCount => throw null;
```
Comment (roji) [inline src/System.Data.Common/ref/System.Data.Common.cs]: As already discussed above, the entire file is full of the older style, without expression-bodied methods/properties. I don't think we should switch to the new style - even for the newly-introduced overloads - to maintain consistency. We can always do a separate PR later to convert the entire file.

File: src/System.Data.Common/ref/System.Data.Common.cs [hunk 6]
```diff
@@ -1939,58 +1939,85 @@ public abstract partial class DbDataReader : System.MarshalByRefObject, System.C
...
+        public virtual System.IO.TextReader GetTextReader(int ordinal) => throw null;
+        public System.IO.TextReader GetTextReader(string name) => throw null;
         public abstract object GetValue(int ordinal);
+        public object GetValue(string name) => throw null;
         public abstract int GetValues(object[] values);
         public abstract bool IsDBNull(int ordinal);
-        public System.Threading.Tasks.Task<bool> IsDBNullAsync(int ordinal) { throw null; }
-        public virtual System.Threading.Tasks.Task<bool> IsDBNullAsync(int ordinal, System.Threading.CancellationToken cancellationToken) { throw null; }
+        public bool IsDBNull(string name) => throw null;
+        public System.Threading.Tasks.Task<bool> IsDBNullAsync(int ordinal) => throw null;
```
Comment (roji) [inline src/System.Data.Common/ref/System.Data.Common.cs]: This has been removed from the implementation but not from here - please remove.

File: src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs [hunk 7]
```diff
@@ -41,12 +40,7 @@ public DbDataReaderMock()
 
         public DbDataReaderMock(DataTable testData)
         {
-            if (testData == null)
-            {
-                throw new ArgumentNullException(nameof(testData));
-            }
-
-            _testDataTable = testData;
+            _testDataTable = testData ?? throw new ArgumentNullException(nameof(testData));
```
Comment (roji) [inline src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs]: I'd rather this PR was constrained to the method overloads we're adding, without modernizing or doing anything beyond that - let's separate API changes from internal cleanup. But it would definitely be good to have these cleanups (as well as some others I'm seeing) in a separate PR after this one.

File: src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs [hunk 8]
```diff
@@ -90,12 +84,21 @@ public override long GetBytes(int ordinal, long dataOffset, byte[] buffer, int b
...
         public override char GetChar(int ordinal)
         {
-            throw new NotImplementedException();
+            return (char)GetValue(ordinal);
         }
 
         public override long GetChars(int ordinal, long dataOffset, char[] buffer, int bufferOffset, int length)
         {
-            throw new NotImplementedException();
+            object value = GetValue(ordinal);
```
Comment (roji) [inline src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs]: This seems to be going beyond what a mock object should be implementing, and I'm not sure it has much value, i.e. just for checking that the string overloads do indeed delegate to the int overloads. But it's no big deal either.

File: src/System.Data.Common/src/System/Data/Common/DbDataReader.cs [hunk 9]
```diff
@@ -103,23 +138,27 @@ public virtual Type GetProviderSpecificFieldType(int ordinal)
...
-        ]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Type GetProviderSpecificFieldType(string name) => GetProviderSpecificFieldType(GetOrdinal(name));
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
         public virtual object GetProviderSpecificValue(int ordinal)
         {
             // NOTE: This is virtual because not all providers may choose to support
-            //       this method, since it was added in Whidbey
+            //       this method, since it was added in Whidbey.
```
Comment (roji) [inline src/System.Data.Common/src/System/Data/Common/DbDataReader.cs]: Nit: please revert unrelated change.

File: src/System.Data.Common/ref/System.Data.Common.cs [hunk 10]
```diff
@@ -1939,58 +1939,85 @@ public abstract partial class DbDataReader : System.MarshalByRefObject, System.C
...
-        public System.Data.Common.DbDataReader GetData(int ordinal) { throw null; }
+        public System.Data.Common.DbDataReader GetData(int ordinal) => throw null;
+        [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
+        public System.Data.Common.DbDataReader GetData(string name) => throw null;
         public abstract string GetDataTypeName(int ordinal);
+        public string GetDataTypeName(string name) => throw null;
         public abstract System.DateTime GetDateTime(int ordinal);
-        protected virtual System.Data.Common.DbDataReader GetDbDataReader(int ordinal) { throw null; }
+        public System.DateTime GetDateTime(string name) => throw null;
+        protected virtual System.Data.Common.DbDataReader GetDbDataReader(int ordinal) => throw null;
```
[the remark this answers is not in the document]
Comment (TheBlueSky) [reply src/System.Data.Common/ref/System.Data.Common.cs]: Which one was removed? `System.DateTime GetDateTime(string name)`? As far as I can see, it's in the implementation and in the final approved API.

File: src/System.Data.Common/ref/System.Data.Common.cs [hunk 11]
```diff
@@ -1939,58 +1939,85 @@ public abstract partial class DbDataReader : System.MarshalByRefObject, System.C
...
         public abstract double GetDouble(int ordinal);
+        public double GetDouble(string name) => throw null;
         [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
         public abstract System.Collections.IEnumerator GetEnumerator();
         public abstract System.Type GetFieldType(int ordinal);
-        public System.Threading.Tasks.Task<T> GetFieldValueAsync<T>(int ordinal) { throw null; }
-        public virtual System.Threading.Tasks.Task<T> GetFieldValueAsync<T>(int ordinal, System.Threading.CancellationToken cancellationToken) { throw null; }
-        public virtual T GetFieldValue<T>(int ordinal) { throw null; }
+        public System.Type GetFieldType(string name) => throw null;
+        public System.Threading.Tasks.Task<T> GetFieldValueAsync<T>(int ordinal) => throw null;
```
[the remark this answers is not in the document]
Comment (TheBlueSky) [reply src/System.Data.Common/ref/System.Data.Common.cs]: Same as the previous similar comment. `public System.Type GetFieldType(string name)` is in the approved API.

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 5 above)
Comment (TheBlueSky) [reply src/System.Data.Common/ref/System.Data.Common.cs]: @roji, what about the new file that holds the new methods; `System.Data.Common.netcoreapp.cs`? Does it need to follow the other file or can it use `=> throw null`?

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 6 above)
Comment (TheBlueSky) [reply src/System.Data.Common/ref/System.Data.Common.cs]: Same as the previous similar comment. `public bool IsDBNull(string name)` is in the approved API.

File: src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs (hunk 8 above)
Comment (TheBlueSky) [reply src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs]: @roji, I didn't quite understand what you meant here. Do you mean testing this change is not needed?

**Comment by TheBlueSky** [comment]
> The fix is to make the ref class partial and use a second file called System.Data.Common.netcoreapp.cs to hold the new parts

@Wraith2, What about `uap` target? In this case, building against `uap` will fail because any of the new methods `does not exist in the reference but it does exist in the implementation`.

**Comment by Wraith2** [comment]
There's a similar msbuild condition for UAP, if you look in the SqlClient project I know it'll be in there somewhere. It'll just be a case of adding the second condition to prevent the new methods being compiled in.

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 5 above)
Comment (Wraith2) [reply src/System.Data.Common/ref/System.Data.Common.cs]: I'd say the new file can be all new syntax because it's a new file it won't hurt readability whien reviewing the changes here. Mixing changes in a PR makes it harder to see which ones relate to specific tasks. You want to try and minimize the change set as much as possible which means usually you end up with small and frequent PR's.

**Comment by TheBlueSky** [comment]
@Wraith2, excuse my ignorance here. If we want to compile these for `netcoreapp` only, shouldn't we have something like `DbDataReader.netcoreapp.cs` file, similar to the `ref` file you suggested before?

**Comment by Wraith2** [comment]
yes, same structure for src and ref, ref just contains the null return/throwing versions as you've already done, src contains functioning implementations.

File: src/System.Data.Common/src/System/Data/Common/DbDataReader.cs (hunk 9 above)
Comment (TheBlueSky) [reply src/System.Data.Common/src/System/Data/Common/DbDataReader.cs]: Reverted.

File: src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs (hunk 7 above)
Comment (TheBlueSky) [reply src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs]: Reverted.

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (roji) [reply src/System.Data.Common/ref/System.Data.Common.cs]: Hmm, it seemed while reviewing that overloads for GetFieldValueAsync, IsDBNullAsync and GetDbDataReader were still there - maybe you fixed and squashed since then. In any case, it doesn't matter, it's looking good now.

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (roji) [reply src/System.Data.Common/ref/System.Data.Common.cs]: As above.

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 5 above)
Comment (roji) [reply src/System.Data.Common/ref/System.Data.Common.cs]: Agree with @Wraith2. In any case the idea would be to some cleanup in a later PR and simplify all of System.Data - it's just better to keep it out of this PR.

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 6 above)
Comment (roji) [reply src/System.Data.Common/ref/System.Data.Common.cs]: As above, the comment was about a new IsDBNullAsync overload which is no longer there.

File: src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs (hunk 8 above)
Comment (roji) [reply src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs]: In general I do think the test value here is a bit limited - but let's leave the code as it is.

Comment (roji) [reply src/System.Data.Common/tests/System/Data/Common/DbDataReaderMock.cs]: As above.

**Comment by roji** [comment]
@TheBlueSky thanks for the quick fixes - to me this looks ready. The remaining test fix seems related to a Helix infrastructure issue (unrelated to your PR).

@terrajobst @divega what's the next step for this?

File: src/System.Data.Common/ref/System.Data.Common.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (TheBlueSky) [reply src/System.Data.Common/ref/System.Data.Common.cs]: Oh, now I see. I think the confusion happened because of the change I did for `=> throw null` which looked as if I added `GetDbDataReader(int ordinal)`. The methods that take an `int` where already there.
Sorry for the confusion.

**Comment by TheBlueSky** [comment]
@roji, thanks for reviewing this PR; I learnt couple of things along the way.

The next step should be someone to tell me why the macOS build times out, and whether this is a big deal or not :)

**Comment by Wraith2** [comment]
The mac build failed on my PR yesterday as well, i'm assuming it's a build server problem and someone in engineering will poke it when they get into the office on US time. Given that you're not doing anything platform specific and the linux builds are working i'd probably ignore it.

**Comment by roji** [comment]
As @Wraith2 wrote, I wouldn't worry about the MacOS build. There's some flakiness in the build infrastructure sometimes.

**Comment by divega** [comment]
@dotnet-bot test OSX x64 Debug Build please

**Comment by safern** [comment]
> @dotnet-bot test OSX x64 Debug Build please

This comment no longer works. I was going to retry the failing leg manually, but accidentally clicked the wrong button and ran all the legs, sorry about that 😢

**Comment by divega** [comment]
> This comment no longer works.

Well, it worked for me, though not directly :trollface:

**Comment by Wraith2** [comment]
Linux failed this time. How do you work out what went wrong? all I see is the send to helix step failed but that doesn't mean anything or provide any point to debug from.

**Comment by safern** [comment]
In the logs you can see the link to test results and a summary of what failed. It seems like infrastructure stuff, trying to download the .zip folder with the test binaries:

```
  Results will be available from https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F35211~2Fmerge/test~2Ffunctional~2Fcli~2F/20190225.17
  Sent Helix Job ef41298b-9ac6-4436-83ba-28a034722d3c
  Waiting for completion of job ef41298b-9ac6-4436-83ba-28a034722d3c
  Job ef41298b-9ac6-4436-83ba-28a034722d3c is completed with 209 finished work items.
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.IO.Compression.Brotli.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.IO.Compression.ZipFile.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.IO.Compression.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item XsltCompiler.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.Xml.Xsl.XslTransformApi.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.Net.Ping.Functional.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code 1 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.Runtime.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code 1 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Job 'ef41298b-9ac6-4436-83ba-28a034722d3c' has 5 failed work items. [/__w/1/s/eng/sendtohelix.proj]

Build FAILED.

/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.IO.Compression.Brotli.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.IO.Compression.ZipFile.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.IO.Compression.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item XsltCompiler.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.Xml.Xsl.XslTransformApi.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code -3 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.Net.Ping.Functional.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code 1 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Work item System.Runtime.Tests in job ef41298b-9ac6-4436-83ba-28a034722d3c has Failed with exit code 1 [/__w/1/s/eng/sendtohelix.proj]
/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19121.4/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : Job 'ef41298b-9ac6-4436-83ba-28a034722d3c' has 5 failed work items. [/__w/1/s/eng/sendtohelix.proj]
```

**Comment by ajcvickers** [comment]
Merging based on sign off from @divega @roji.

**Comment by divega** [comment]
Thank you @TheBlueSky for the contribution and @Wraith2, @safern for all the help getting it in.

@roji is this the first of all the ADO.NET improvements you filed to get fixed? :smile:

**Comment by TheBlueSky** [comment]
Yay! :)

Thanks for everyone who helped in this, especially @Wraith2.

**Comment by roji** [comment]
> @roji is this the first of all the ADO.NET improvements you filed to get fixed? 😄

Yeah, I think it is, after all these years :) A sign of more PRs to come :)
